IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m

```diff
@@ -556,4 +560,183 @@ - (BOOL)removeAppMetadata:(nonnull MSIDAppMetadataCacheItem *)appMetadata
+                    }
+                }
+                break;
+                
+            // FRT is disabled, check to see if should be enabled
+            case MSIDIsFRTEnabledStatusDisabledByKeychainItem:
+                if (shouldEnableFRT)
+                {
+                    [self updateFRTSettings:YES context:context error:&updateError];
+                    newStatus = MSIDIsFRTEnabledStatusEnabled;
+                }
+                break;
+                
+                // Error reading keychain item, do not update settings
+            case MSIDIsFRTEnabledStatusDisabledByKeychainError:
+                break;
+                
+                // Feature is disabled by client app, do nothing with keychain item
+            case MSIDIsFRTEnabledStatusDisabledByClientApp:
+                break;
+        }
+        
+        if (updateError)
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Error when trying to update FRT settings, error: %@", updateError);
+            newStatus = status;
+        }
+        
+        return newStatus;
+    };
+    
+    // Check if FRT is disabled by client
+    if (context.disableFRT)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"FRT disabled by MSAL client app, returning NO");
+        return checkFeatureFlagsAndReturn(MSIDIsFRTEnabledStatusDisabledByClientApp);
+    }
+    
+    NSError *readError = nil;
+    NSArray<MSIDJsonObject *> *jsonObjects = [_dataSource jsonObjectsWithKey:[MSIDAccountCredentialCache checkFRTCacheKey]
+                                                                  serializer:[MSIDCacheItemJsonSerializer new]
+                                                                     context:context
+                                                                       error:&readError];
+    
+    if (readError)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Failed to retrieve FRT cache entry, error: %@", readError);
+        if (error)
+        {
+            *error = readError;
+        }
+        return checkFeatureFlagsAndReturn(MSIDIsFRTEnabledStatusDisabledByKeychainError);
+    }
+    
+    if (![jsonObjects count])
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"No FRT cache entry found, returning NO");
+        return checkFeatureFlagsAndReturn(MSIDIsFRTEnabledStatusNotEnabled);
+    }
+    
+    NSDictionary *dict = [jsonObjects[0] jsonDictionary];
+    if (!dict || ![dict isKindOfClass:[NSDictionary class]] || [dict objectForKey:MSID_USE_SINGLE_FRT_KEY] == nil)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Failed to deserialize FRT cache entry, returning NO");
+        return checkFeatureFlagsAndReturn(MSIDIsFRTEnabledStatusDisabledByDeserializationError);
+    }
+    
+    id useFRT = dict[MSID_USE_SINGLE_FRT_KEY];
+    if(([useFRT isKindOfClass:[NSNumber class]] || [useFRT isKindOfClass:[NSString class]]) && [useFRT boolValue])
```

Comment:
we have "msidAssertType" method in NSDictionary category to check keys types in the json dict.

Reply:
Updated